File: Sorting/MergeSort.java
public class MergeSort
{
	public static int[] merge_sort(int arr[])
	{
		int arr_length = arr.length;

		if(arr_length <2) //if array is too short then doing nothing
		{
			return arr;
		}	
	
		int mid = (arr_length)/2;
		int left[] = new int[mid];
		int right[] = new int[arr_length-mid];

		for(int i=0; i<mid; i++)
		{
			left[i] = arr[i];
		}

		for(int i=mid; i<arr_length; i++)
		{
			right[i-mid] = arr[i];
		}

		merge_sort(left);
		merge_sort(right);
		merge(left, right, arr);			

		return arr;
	}

	public static void merge(int left[], int right[], int arr[])
	{
		int left_length = left.length;
		int right_length = right.length;
		int arr_length = arr.length;

		int i=0, j=0, k=0;

		while(i<left_length && j<right_length)
		{
			if(left[i]<=right[j])
			{
				arr[k] = left[i];
				i++;
			}
			else
			{
				arr[k] = right[j];
				j++;
			}

			k++;
		}

		while(i< left_length)
		{
			arr[k] = left[i];
			i++;
			k++;
		}

		while(j< right_length)
		{
			arr[k] = right[j];
			j++;
			k++;
		}
	}

	public static void main(String args[])
	{
	//creating and printing the initial array
		int arr[] = {1,9,5,0,4,8,11,89,6,7,500,3,13};

        for(int k: arr)
        {
            System.out.print(k + ", ");
        }
        System.out.println("");

    //sorting the array and displaying it
        int arr2[] = merge_sort(arr);

        for(int k: arr2)
        {
            System.out.print(k + ", ");
        }
        
	}
}

File: Sorting/MergeSort.py
def mergeSort(arr):

	length = int(len(arr))
	mid= int(len(arr)/2)

	if length <2:
		return arr

	left = []
	right = []

	for i in range(mid):
		left.append(arr[i])

	for j in range(length-mid):
		right.append(arr[mid+j])

	# print(left)
	# print(right)

	mergeSort(left)
	mergeSort(right)
	merge(left, right, arr)

	return arr


def merge(left, right, arr):
	i=0
	j=0
	k=0

	left_len = len(left)
	right_len = len(right)
	arr_len = len(arr)

	while i< left_len and j< right_len:
		if left[i] < right[j]:
			arr[k] = left[i]
			i +=1
		elif left[i] > right[j]:
			arr[k] = right[j]
			j +=1
		
		k +=1

	while i<left_len:
		arr[k] = left[i]
		i +=1
		k +=1

	while j<right_len:
		arr[k] = right[j]
		j +=1
		k +=1

arr = [8,5,9,7,6,2,4,1,3]

mergeSort(arr)
print(arr)